install_scripts/templates/common/kubernetes-upgrade.sh

```diff
@@ -135,12 +137,217 @@ maybeUpgradeKubernetes() {
+#######################################
+isLoadBalancerAddressChanging() {
+    local loadBalancerAddress="$1"
+    if [ -z "$loadBalancerAddress" ] || ! kubeadm config view >/dev/null 2>&1; then
+        return 1
+    fi
+    local previous_address="$(kubeadm config view | grep 'controlPlaneEndpoint:' | sed 's/controlPlaneEndpoint: \|"//g')"
+    if [ -z "$previous_address" ]; then
+        return 1
+    fi
+    splitHostPort "$previous_address"
+    local previous_address="$HOST"
+    local previous_port="$PORT"
+    if [ -z "$previous_port" ]; then
+        previous_port=6443
+    fi
+    splitHostPort "$loadBalancerAddress"
+    local next_address="$HOST"
+    local next_port="$PORT"
+    if [ -z "$next_port" ]; then
+        next_port=6443
+    fi
+    if [ "$previous_address" = "$next_address" ] && [ "$previous_port" = "$next_port" ]; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+#######################################
+# TODO
+# Globals:
+#   LOAD_BALANCER_ADDRESS
+#   LOAD_BALANCER_PORT
+# Arguments:
+#   k8sTargetVersion - e.g. 1.10.6
+# Returns:
+#   DID_UPGRADE_KUBERNETES
+#######################################
+DID_UPGRADE_KUBERNETES=0
+maybeUpgradeKubernetesLoadBalancer() {
+    if [ -z "$LOAD_BALANCER_ADDRESS" ] || [ -z "$LOAD_BALANCER_PORT" ]; then
+        return
+    fi
+    if ! isLoadBalancerAddressChanging "$LOAD_BALANCER_ADDRESS:$LOAD_BALANCER_PORT"; then
+        updateKubernetesAPIServerCerts "$LOAD_BALANCER_ADDRESS" "$LOAD_BALANCER_PORT"
+        updateKubeconfigs "https://$LOAD_BALANCER_ADDRESS:$LOAD_BALANCER_PORT"
+        return
+    fi
+
+    local k8sTargetVersion="$1"
+    semverParse "$k8sTargetVersion"
+    local k8sTargetMajor="$major"
+    local k8sTargetMinor="$minor"
+    local k8sTargetPatch="$patch"
+
+    if [ "$k8sTargetMinor" -lt 13 ]; then
+        return
+    fi
+
+    echo ""
+    logStep "Kubernetes control plane endpoint updated, upgrading control plane..."
+    echo ""
+
+    # this will use the new load balancer address to the kubeadm.conf
+    : > /opt/replicated/kubeadm.conf
+    makeKubeadmConfig
+
+    updateKubernetesAPIServerCerts "$LOAD_BALANCER_ADDRESS" "$LOAD_BALANCER_PORT"
+    updateKubeconfigs "https://$LOAD_BALANCER_ADDRESS:$LOAD_BALANCER_PORT"
+
+    logStep "Upgrading kubernetes control plane"
+    kubeadm upgrade apply -yf --config=/opt/replicated/kubeadm.conf
+    logSuccess "Kubernetes control plane upgraded"
+
+    logStep "Restarting kube-proxy"
+    kubectl -n kube-system get pods | grep kube-proxy | awk '{print $1}' | xargs kubectl -n kube-system delete pod
+    logSuccess "Kube-proxy restarted"
+
+    spinnerNodesReady
+
+    if [ "$(kubectl get nodes | grep master | wc -l)" -gt "1" ]; then
+        echo ""
+        printf "Run the upgrade script on remote master nodes before proceeding:\n\n${GREEN}"
+        replicatedctl cluster node-join-script --master | sed "s/api-service-address=[^ ]*/api-service-address=$LOAD_BALANCER_ADDRESS:$LOAD_BALANCER_PORT/"
```

Comment:
Doesn’t this need to download the pki bundle? Does it?

Reply:
It was regenerating the apiserver.crt on the other nodes using the previously downloaded CA. Does it need to be the same cert across the cluster?